### PR TITLE
Support axes argument of gufuncs for Masked inputs

### PR DESCRIPTION
**Pull request by mhvk**
This pull request builds on #16120 to add support for `axes` for gufuncs with `Masked` inputs or outputs. It also refactors `__array_ufunc__` a little, ensuring that if `Masked` output is passed in, its mask always get used for possible in-place mask calculations.

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I think we should fix #16123 before merging this

**Comment by mhvk** [comment]
> I think we should fix #16123 before merging this

I think it is somewhat orthogonal to what goes wrong in #16123, so it would seem fine to deal with this PR independently (#16123 will require changes to the `__call__` path, which this PR barely touches, so it won't be much trouble to rebase/backport).

**Comment by neutrinoceros** [comment]
Ok then. I can review this one after https://github.com/astropy/astropy/pull/16120 goes in !

**Comment by pllim** [comment]
#16120 is merged.

File: astropy/utils/masked/core.py
Comment (neutrinoceros) [inline astropy/utils/masked/core.py:833]: just curious, why not use `np.copyto` here ? is this significantly faster ?

**Review by neutrinoceros — COMMENTED** [review]
Looks good, mostly minor questions (and nits) !

File: astropy/utils/masked/tests/test_masked.py
Comment (mhvk) [reply astropy/utils/masked/tests/test_masked.py:761]: Good idea!

File: astropy/utils/masked/tests/test_masked.py
Comment (mhvk) [reply astropy/utils/masked/tests/test_masked.py:763]: It is meant to be the size of the dimension that is operated on; I've replaced with `m1.shape[axes[0][1]]` (which will at least hint at what this is).

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:693]: You're right, this is inconsistent. Changed...

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:754]: Done!

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:755]: Done!

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:784]: One can only give one or the other, and by moving the actual unmasked call to before the whole mask-dealing bit, it has been checked. I'll add a note.

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:787]: `axes` will have been checked already to have the right length.

Comment (mhvk) [reply astropy/utils/masked/core.py:787]: Well, that's not completely true: it will have at least `nin` elements, but `nout` can be omitted if they're all `()`. I'll adjust!

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:785]: thanks, indeed!

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:833]: Ah, I remember now, `np.copyto` does broadcasting, so when one also has a `where` argument, one has no choice.

File: astropy/utils/masked/core.py
Comment (mhvk) [reply astropy/utils/masked/core.py:807]: This one was indeed right. Now added test...

**Review by neutrinoceros — APPROVED** [review]
All good now, thanks for walking me through this !

**Review by pllim — APPROVED** [review]
Again, I am approving by proxy to unblock merge. I'll let mhvk decide when to merge.

Thanks!

**Comment by mhvk** [comment]
Thanks, @pllim and @neutrinoceros!

**Comment by pllim** [comment]
You did the hard work. Thank YOU!

**Pull request merged**
